pkg/controller/dataimportcron-controller.go

```diff
@@ -1091,6 +1098,90 @@ func addDataImportCronControllerWatches(mgr manager.Manager, c controller.Contro
 	return nil
 }
 
+// addDefaultStorageClassUpdateWatch watches for default/virt default storage class updates
+func addDefaultStorageClassUpdateWatch(mgr manager.Manager, c controller.Controller) error {
+	if err := c.Watch(source.Kind(mgr.GetCache(), &storagev1.StorageClass{}),
+		handler.EnqueueRequestsFromMapFunc(
+			func(ctx context.Context, obj client.Object) []reconcile.Request {
+				log := c.GetLogger().WithName("DefaultStorageClassUpdateWatch")
+				log.Info("Update", "sc", obj.GetName(),
+					"default", obj.GetAnnotations()[cc.AnnDefaultStorageClass] == "true",
+					"defaultVirt", obj.GetAnnotations()[cc.AnnDefaultVirtStorageClass] == "true")
+				reqs, err := getReconcileRequestsForDicsWithPendingPvc(ctx, mgr.GetClient())
+				if err != nil {
+					log.Error(err, "Failed getting DataImportCrons with pending PVCs")
+				}
+				return reqs
+			},
+		),
+		predicate.Funcs{
+			CreateFunc: func(event.CreateEvent) bool { return false },
+			DeleteFunc: func(event.DeleteEvent) bool { return false },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				scOld := e.ObjectOld.(*storagev1.StorageClass)
+				scNew := e.ObjectNew.(*storagev1.StorageClass)
+				return (scNew.Annotations[cc.AnnDefaultStorageClass] != scOld.Annotations[cc.AnnDefaultStorageClass]) ||
+					(scNew.Annotations[cc.AnnDefaultVirtStorageClass] != scOld.Annotations[cc.AnnDefaultVirtStorageClass])
+			},
+		},
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getReconcileRequestsForDicsWithPendingPvc(ctx context.Context, c client.Client) ([]reconcile.Request, error) {
+	dicList := &cdiv1.DataImportCronList{}
+	if err := c.List(ctx, dicList); err != nil {
+		return nil, err
+	}
+	reqs := []reconcile.Request{}
+	for _, dic := range dicList.Items {
+		if cc.GetStorageClassFromDVSpec(&dic.Spec.Template) != nil {
+			continue
+		}
+
+		imports := dic.Status.CurrentImports
+		if len(imports) == 0 {
+			continue
+		}
+
+		pvcName := imports[0].DataVolumeName
+		pvc := &corev1.PersistentVolumeClaim{}
+		if err := c.Get(ctx, types.NamespacedName{Namespace: dic.Namespace, Name: pvcName}, pvc); err != nil {
+			if k8serrors.IsNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+
+		if pvc.Status.Phase == corev1.ClaimPending {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Name: dic.Name, Namespace: dic.Namespace}})
+		}
+	}
+
+	return reqs, nil
+}
+
+func (r *DataImportCronReconciler) deleteOutdatedPendingPvc(ctx context.Context, pvc *corev1.PersistentVolumeClaim, desiredStorageClass, cronName string) (bool, error) {
+	if pvc == nil || pvc.Status.Phase != corev1.ClaimPending || pvc.Labels[common.DataImportCronLabel] != cronName {
+		return false, nil
+	}
+
+	sc := pvc.Spec.StorageClassName
+	if sc == nil || *sc == desiredStorageClass {
+		return false, nil
+	}
```

Comment:
Doesn't this result in an infinite loop of deleting PVCs? maybe paranoid but we use informers for storage class get so temporary lying is possible/unsure when pvc.Spec.StorageClassName gets populated

Reply:
Not sure I see how can it result an infinite loop of deleting PVCs. If the pvc.Spec.StorageClassName is not populated yet we shouldn't delete the PVC.

When we are reconciled due to the SC update watch, aren't we guaranteed the SCs we get in the reconcile are updated? otherwise how can we build on watching informer-ized CRDs in general?

Reply:
Right, since you're watching StorageClass objects you are guaranteed to have the latest storage class in the reconcile loop